source/grid/grid_tools.cc

```diff
@@ -1501,6 +1502,105 @@ namespace GridTools
+            return true;
+        }
+      return false;
+    }
+
+  }
+
+
+
+  template <class Container>
+  std::vector<typename Container::active_cell_iterator>
+  compute_active_cell_halo_layer (const Container                                                                    &container,
+                                  const std_cxx11::function<bool (const typename Container::active_cell_iterator &)> &predicate)
+  {
+    std::vector<typename Container::active_cell_iterator> active_halo_layer;
+    std::vector<bool> locally_active_vertices_on_subdomain (get_tria(container).n_vertices(),
+                                                            false);
+
+    // Find the cells for which the predicate is true
+    // These are the cells around which we wish to construct
+    // the halo layer
+    for (typename Container::active_cell_iterator
+         cell = container.begin_active();
+         cell != container.end(); ++cell)
+      if (predicate(cell)) // True predicate --> Part of subdomain
+        for (unsigned int v=0; v<GeometryInfo<Container::dimension>::vertices_per_cell; ++v)
+          locally_active_vertices_on_subdomain[cell->vertex_index(v)] = true;
+
+    // Find the cells that do not conform to the predicate
+    // but share a vertex with the selected subdomain
+    // These comprise the halo layer
+    for (typename Container::active_cell_iterator
+         cell = container.begin_active();
+         cell != container.end(); ++cell)
+      if (!predicate(cell)) // False predicate --> Potential halo cell
+        for (unsigned int v=0; v<GeometryInfo<Container::dimension>::vertices_per_cell; ++v)
+          if (locally_active_vertices_on_subdomain[cell->vertex_index(v)] == true)
+            {
+              active_halo_layer.push_back(cell);
+              break;
+            }
+
+    // Check that we never return artificial cells
+    // Its alright for us to return ghost cells though
+    Assert(contains_artificial_cells<Container>(active_halo_layer) == false,
+           ExcMessage("Halo layer contains artificial cells"));
+
+    return active_halo_layer;
+  }
+
+  template <class Container>
+  std::vector<typename Container::active_cell_iterator>
+  compute_ghost_cell_halo_layer (const Container &container)
+  {
+    return compute_active_cell_halo_layer (container, IteratorFilters::LocallyOwnedCell());
+  }
+
+
+#ifdef DEAL_II_WITH_CXX11
+
+
+  template <class Container>
+  std::vector<typename Container::active_cell_iterator>
+  compute_active_cell_halo_layer (const Container                    &container,
+                                  const std::set<types::material_id> &materials)
+  {
+    const std_cxx11::function<bool (const typename Container::active_cell_iterator &)>
+    predicate ([&](const typename Container::active_cell_iterator & cell)
+    {
+      return cell->is_locally_owned() &&
+             materials.find(cell->material_id()) != materials.end();
+    });
```

Comment:
I think you could just define this function in an anonymous namespace just in front of this instance of `compute_active_cell_halo_layer` and then enable this and the next function also in non-C++11 mode.